apollo-router/src/compute_job/metrics.rs

```diff
@@ -0,0 +1,188 @@
+use std::time::Duration;
+use std::time::Instant;
+
+use crate::compute_job::ComputeJobType;
+use crate::plugins::telemetry::consts::OTEL_STATUS_CODE_ERROR;
+use crate::plugins::telemetry::consts::OTEL_STATUS_CODE_OK;
+
+#[derive(Copy, Clone, strum_macros::IntoStaticStr)]
+pub(super) enum Outcome {
+    Executed,
+    ExecutedError,
+    RejectedQueueFull,
+    Abandoned,
+}
+
+impl Outcome {
+    fn as_otel_status(&self) -> &'static str {
+        match self {
+            Self::Executed => OTEL_STATUS_CODE_OK,
+            Self::Abandoned | Self::ExecutedError | Self::RejectedQueueFull => {
+                OTEL_STATUS_CODE_ERROR
+            }
+        }
+    }
+}
+
+impl From<Outcome> for opentelemetry::Value {
+    fn from(outcome: Outcome) -> Self {
+        let s: &'static str = outcome.into();
+        s.into()
+    }
+}
+
+pub(super) struct JobWatcher {
+    queue_start: Instant,
+    compute_job_type: ComputeJobType,
+    pub(super) outcome: Outcome,
+}
+
+impl JobWatcher {
+    pub(super) fn new(compute_job_type: ComputeJobType) -> Self {
+        Self {
+            queue_start: Instant::now(),
+            outcome: Outcome::Abandoned,
+            compute_job_type,
+        }
+    }
+}
+
+impl Drop for JobWatcher {
+    fn drop(&mut self) {
+        let current_span = tracing::Span::current();
+        current_span.record::<str, &'static str>("job.outcome", self.outcome.into());
+        current_span.record("otel.status_code", self.outcome.as_otel_status());
```

Comment:
These are for updating the span which is not part of this PR.

Reply:
[2a35433](https://github.com/apollographql/router/pull/7184/commits/2a35433fe76b3c52298430579f947980f3410b29)